mathjax3-ts/input/tex/ParseUtil.ts

```diff
@@ -488,6 +488,110 @@ namespace ParseUtil {
     return (font ? {mathvariant: font} : {});
   };
 
+
+  /**
+   * Splits a package option list of the form [x=y,z=1] into an attribute list
+   * of the form {x: y, z: 1}.
+   * @param {string} attrib The attributes of the package.
+   * @param {{[key: string]: number}?} allowed A list of allowed options. If
+   *     given only allowed arguments are returned.
+   * @param {boolean?} error If true, raises an exception if not allowed options
+   *     are found.
+   * @return {EnvList} The attribute list.
+   */
+  export function keyvalOptions(attrib: string,
+                                allowed: {[key: string]: number} = null,
+                                error: boolean = false): EnvList {
+    let def: EnvList = readKeyval(attrib);
+    if (allowed) {
+      for (let key of Object.keys(def)) {
+        if (!allowed[key]) {
+          if (error) {
+            throw new TexError('InvalidOption',
+                               'Invalid optional argument: ', key);
+          }
+          delete def[key];
+        }
+      }
+    }
+    return def;
+  }
+
+
+  /**
+   * Implementation of the keyval function from https://www.ctan.org/pkg/keyval
+   * @param {string} text The optional parameter string for a package or
+   *     command.
+   * @return {EnvList} Set of options as key/value pairs.
+   */
+  function readKeyval(text: string): EnvList {
+    let options: EnvList = {};
+    let rest = text;
+    let end, key, val;
+    while (rest) {
+      [key, end, rest] = readValue(rest, ['=', ',']);
+      if (end === '=') {
+        [val, end, rest] = readValue(rest, [',']);
+        val = ((val === 'false' || val === 'true') ?
+               JSON.parse : removeBraces)(val);
+        options[key] = val;
+      } else if (key) {
+        options[key] = true;
+      }
+    }
+    return options;
+  }
+
+
+  /**
+   * Removes pairs of outer braces.
+   * @param {string} text The string to clean.
+   * @return {string} The cleaned string.
+   */
+  function removeBraces(text: string): string {
+    while (text.match(/^{.*}$/)) {
+      text = text.slice(1, text.length - 1);
+    }
+    return text;
+  }
+
+
+  /**
+   * Read a value from the given string until an end parameter is reached or
+   * string is exhausted.
+   * @param {string} text The string to process.
+   * @param {string[]} end List of possible end characters.
+   * @return {[string, string, string]} The collected value, the actual end
+   *     character, and the rest of the string still to parse.
+   */
+  function readValue(text: string, end: string[]): [string, string, string] {
+    let rest = text;
+    let braces = 0;
+    let value = '';
+    while (rest) {
+      let c = rest[0];
+      rest = rest.slice(1);
```

Comment:
Making all these new copies off the string (one for each character) seems a bit inefficient.  Can you just use an index to move through the string rather than creating new strings each time you move to a new character?

Reply:
Uses `index` counter now.